### PR TITLE
#7774 - Automatic & manual archiving for all core entities

### LINKED ISSUE #7774: Automatic & manual archiving for all core entities [8]
### Feature Description
Currently SORMAS comes with automatic archiving for **cases** and **events**. Those are archived when they were not modified for a certain number of days (90 days default). 
**Contacts** are currently not archived on their own. Instead they use the archived status of the related case. 
**Event participants** also don't make use of archiving, because they depend on the related event.
**Immunizations** and **travel entries** are currently not archived at all.
The configuration for automatic archiving is implemented in #7775

### Problem Description

1. Contacts that are not related to a case currently aren't achived
2. It's currently not possible to archive a contact independent from it's case
3. Automatic archiving of events does not consider changes made to event participants
4. Automatic archiving of cases and events does not consider changes made to related samples
5. Event participants can't be archived on their own
6. Contact, event participants, immunizations and travel entries are curently not automatically archived.

### Proposed Change

Changes related to manual archiving:

- [x] Likely already done, but should be checked again: Enable archiving of contacts and event participants. Add the needed database column (currently the property is transient). 
Pre-fill the value based on the archived status of the related case / event.
- [x] Adjust the methods that provide data for the UI directory views and the mobile sync accordingly: Contacts should no longer consider the archived status of the related case. Event participants should consider the status of the event in addition to their own
- [x] Allow manual archiving of contacts in the UI. Add the needed UserRight.
- [x] Allow manual archiving of event participants in the UI. Add the needed UserRight
- [x] When a case is manually archived, ask the user whether all contacts should also be archived. Add a separate archiveWithContacts method to the CaseFacade for the archiving logic. 
- [x] The same has to be implemented for bulk (de-)archiving of cases
- [x] When an event is manually or automatically archived also archive all event participants related to it. This should be implemented in EventFacade.archive
- [x] The same has to be implemented for bulk (de-)archiving of events
- [x] When manually archiving an entity the end of process date should no longer be editable. Just show it in an read-only field. Note: The date should be based on the aggregated change date of the entity. This has already been implemented by #7319

Changes related to automatic archiving:

- [x] When automatically archiving cases, also consider the change date of related samples, surveillance reports and visits (**not** contacts and not the visits of those contacts)
- [x] When automatically archiving events also consider the change date of event actions, event participants and their samples
- [x] When an event is automatically archived also archive all event participants of the event
- [x] Implement automatic archiving for contacts. For the date logic also consider related samples and visits
- [x] Implement automatic archiving for event participants. For the date logic also consider related samples
- [x] Implement automatic archiving for immunizations. For the date logic also consider related vaccinations
- [x] Implement automatic archiving for travel entries

**Important**: We have run into performance problems with automatic archiving before (see #2894). Thus automatic archiving needs to be done in bulk.

Documentation (this should probably be done by @MartinWahnschaffe)

- [ ] Create/extend the readme to explain in detail how automatic archiving works

### Additional Information

I have separated #7775 from this. If this issue is still too big, we could separate the work related to case & contacts into one issue and the other entites into a separate one.

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #7774

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java [hunk 1]
```diff
@@ -25,4 +25,6 @@
 	List<TravelEntryListEntryDto> getEntriesList(TravelEntryListCriteria criteria, Integer first, Integer max);
 
 	Page<TravelEntryIndexDto> getIndexPage(TravelEntryCriteria criteria, Integer first, Integer max, List<SortProperty> sortProperties);
+
+	void archiveAllArchivableImmunizations(int daysAfterTravelEntryGetsArchived);
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java:29]:
```suggestion
	void archiveAllArchivableTravelEntries(int daysAfterTravelEntryGetsArchived);
```

File: sormas-app/app/src/main/java/de/symeda/sormas/app/rest/SynchronizeDataAsync.java [hunk 2]
```diff
@@ -125,7 +126,7 @@ protected Void doInBackground(Void... params) {
 				}
 				// Pull and remove archived entities when the last time this has been done is more than 24 hours ago
 				if (ConfigProvider.getLastArchivedSyncDate() == null
-					|| DateHelper.getFullDaysBetween(ConfigProvider.getLastArchivedSyncDate(), new Date()) >= 1) {
+					|| DateHelper.getFullDaysBetween(ConfigProvider.getLastArchivedSyncDate(), new Date()) >= 0) {
```
Comment (MateStrysewske) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/rest/SynchronizeDataAsync.java:129]: Revert this change (probably for testing purposes)

File: sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java [hunk 3]
```diff
@@ -2415,20 +2415,32 @@ public void deleteCaseAsDuplicate(String caseUuid, String duplicateOfCaseUuid) t
 
 	@Override
 	@RolesAllowed(UserRight._CASE_ARCHIVE)
-	public void archive(String entityUuid, Date endOfProcessingDate) {
+	public void archive(String entityUuid, Date endOfProcessingDate, boolean includeContacts) {
 		super.archive(entityUuid, endOfProcessingDate);
+		if (includeContacts) {
+			List<String> caseContacts = contactService.getAllUuidsByCaseUuids(Collections.singletonList(entityUuid));
+			contactService.archive(caseContacts);
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java:2422]: Shouldn't the end of processing date be set here as well?

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java (hunk 1 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java:29]: changed

File: sormas-app/app/src/main/java/de/symeda/sormas/app/rest/SynchronizeDataAsync.java (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-app/app/src/main/java/de/symeda/sormas/app/rest/SynchronizeDataAsync.java:129]: changed, sorry for that. I thought I did but did not check before finishing

File: sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java (hunk 3 above)
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/caze/CaseFacadeEjb.java:2422]: no, for bulk archiving the endOfProcessingDate is calculated in the archive method from AbstractCoreAdoService

File: sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java [hunk 4]
```diff
@@ -1413,6 +1415,38 @@ public boolean isDeleted(String contactUuid) {
...
+	private void archiveAllArchivableContacts(int daysAfterEventGetsArchived, @NotNull LocalDate referenceDate) {
+		LocalDate notChangedSince = referenceDate.minusDays(daysAfterEventGetsArchived);
+
+		CriteriaBuilder cb = em.getCriteriaBuilder();
+		CriteriaQuery<String> cq = cb.createQuery(String.class);
+		Root<Contact> from = cq.from(Contact.class);
+
+		Timestamp notChangedTimestamp = Timestamp.valueOf(notChangedSince.atStartOfDay());
+		cq.where(cb.equal(from.get(Event.ARCHIVED), false), cb.not(service.createChangeDateFilter(cb, from, notChangedTimestamp)));
+		cq.select(from.get(Event.UUID)).distinct(true);
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/contact/ContactFacadeEjb.java:1442]: changed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/travelentry/TravelEntryFacadeEjb.java [hunk 5]
```diff
@@ -179,6 +185,29 @@ protected void restorePseudonymizedDto(TravelEntryDto dto, TravelEntryDto existi
...
+		LocalDate notChangedSince = referenceDate.minusDays(daysAfterTravelEntryGetsArchived);
+
+		CriteriaBuilder cb = em.getCriteriaBuilder();
+		CriteriaQuery<String> cq = cb.createQuery(String.class);
+		Root<TravelEntry> from = cq.from(TravelEntry.class);
+
+		Timestamp notChangedTimestamp = Timestamp.valueOf(notChangedSince.atStartOfDay());
+		cq.where(cb.equal(from.get(Event.ARCHIVED), false), cb.not(service.createChangeDateFilter(cb, from, notChangedTimestamp)));
+		cq.select(from.get(TravelEntry.UUID)).distinct(true);
+		List<String> immunizationUuids = em.createQuery(cq).getResultList();
```
[the remark this answers is not in the document]
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/travelentry/TravelEntryFacadeEjb.java:204]: changed

File: sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java [hunk 6]
```diff
@@ -25,4 +25,6 @@
 	List<TravelEntryListEntryDto> getEntriesList(TravelEntryListCriteria criteria, Integer first, Integer max);
 
 	Page<TravelEntryIndexDto> getIndexPage(TravelEntryCriteria criteria, Integer first, Integer max, List<SortProperty> sortProperties);
+
+	void archiveAllArchivableTravelEntry(int daysAfterTravelEntryGetsArchived);
```
Comment (MateStrysewske) [inline sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java:29]:
```suggestion
	void archiveAllArchivableTravelEntries(int daysAfterTravelEntryGetsArchived);
```

File: sormas-backend/src/main/java/de/symeda/sormas/backend/common/CronService.java [hunk 7]
```diff
@@ -220,4 +226,45 @@ public void deleteExpiredEntities() {
...
+		final int daysAfterImmunizationsGetsArchived = featureConfigurationFacade
+			.getProperty(FeatureType.AUTOMATIC_ARCHIVING, CoreEntityType.IMMUNIZATION, FeatureTypeProperty.THRESHOLD_IN_DAYS, Integer.class);
+
+		if (daysAfterImmunizationsGetsArchived >= 1) {
+			immunizationFacade.archiveAllArchivableImmunizations(daysAfterImmunizationsGetsArchived);
+		}
+	}
+
+	@Schedule(hour = "2", minute = "30", persistent = false)
+	public void archiveTravelEntry() {
```
Comment (MateStrysewske) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/common/CronService.java:261]:
```suggestion
	public void archiveTravelEntries() {
```

File: sormas-backend/src/main/resources/sql/sormas_schema.sql [hunk 8]
```diff
@@ -10548,4 +10548,17 @@ ALTER TABLE travelentry ALTER COLUMN dateofarrival SET NOT NULL;
 
 INSERT INTO schema_version (version_number, comment) VALUES (449, 'Add dateOfArrival to travel entries #7845');
 
+-- 2022-03-14 - #7774 Automatic & manual archiving for all core entities
+UPDATE contact SET archived = TRUE
```
Comment (MateStrysewske) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:10552]: Please double check these update scripts, especially concerning contacts that don't have a source case; looks like for those, archived is also set to TRUE which definitely shouldn't happen.

**Review by MateStrysewske — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java (hunk 6 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/java/de/symeda/sormas/api/travelentry/TravelEntryFacade.java:29]: changed

File: sormas-backend/src/main/java/de/symeda/sormas/backend/common/CronService.java (hunk 7 above)
Comment (sergiupacurariu) [reply sormas-backend/src/main/java/de/symeda/sormas/backend/common/CronService.java:261]: changed

File: sormas-backend/src/main/resources/sql/sormas_schema.sql (hunk 8 above)
Comment (sergiupacurariu) [reply sormas-backend/src/main/resources/sql/sormas_schema.sql:10552]: changed

**Review by MateStrysewske — APPROVED** [review]

**Comment by MateStrysewske** [comment]
LGTM. I don't know whether @MartinWahnschaffe also wants to have a look at this; if not, feel free to merge it, @sergiupacurariu :)

File: sormas-api/src/main/java/de/symeda/sormas/api/event/EventParticipantFacade.java [hunk 9]
```diff
@@ -66,6 +66,10 @@
 
 	List<EventParticipantDto> getAllActiveEventParticipantsAfter(Date date, Integer batchSize, String lastSynchronizedUuid);
 
+	List<String> getArchivedUuidsSince(Date since);
+
+	void archiveAllArchivableEventParticipants(int daysAfterEventParticipantGetsArchived);
```
Comment (MartinWahnschaffe) [inline sormas-api/src/main/java/de/symeda/sormas/api/event/EventParticipantFacade.java:71]: THis method should not be part of the facade interface (to the outside). It should just be a public method of the facadeEjb that can be called from other backend classes - in this case the CronService.
Same for ImmunizationFacade, etc.

**Comment by MartinWahnschaffe** [comment]
Just the one point I mentioned above. Besides that it can be merged.

File: sormas-api/src/main/java/de/symeda/sormas/api/event/EventParticipantFacade.java (hunk 9 above)
Comment (sergiupacurariu) [reply sormas-api/src/main/java/de/symeda/sormas/api/event/EventParticipantFacade.java]: removed the method from interface
